server/src/main/java/io/crate/execution/engine/aggregation/impl/average/AverageAggregation.java

```diff
@@ -58,148 +55,58 @@
-                return false;
+            for (var supportedType : SUPPORTED_FRACTIONAL_TYPES) {
+                mod.register(
+                    Signature.aggregate(
+                        functionName,
+                        supportedType.getTypeSignature(),
+                        DataTypes.DOUBLE.getTypeSignature()),
+                    (signature, boundSignature) ->
+                        new AverageAggregation(signature, boundSignature, () -> new FractionalAverageState(), () -> FractionalAverageStateType.INSTANCE)
+                );
             }
-            AverageState that = (AverageState) o;
-            return Objects.equals(that.value(), value());
-        }
-
-        @Override
-        public int hashCode() {
-            return Objects.hash(value());
-        }
-    }
-
-    public static class AverageStateType extends DataType<AverageState>
-        implements FixedWidthType, Streamer<AverageState> {
-
-        public static final int ID = 1024;
-        private static final AverageStateType INSTANCE = new AverageStateType();
-        private static final int AVERAGE_STATE_SIZE = (int) RamUsageEstimator.shallowSizeOfInstance(AverageState.class);
-
-        @Override
-        public int id() {
-            return ID;
-        }
-
-        @Override
-        public Precedence precedence() {
-            return Precedence.CUSTOM;
-        }
-
-        @Override
-        public String getName() {
-            return "average_state";
-        }
-
-        @Override
-        public Streamer<AverageState> streamer() {
-            return this;
-        }
-
-        @Override
-        public AverageState sanitizeValue(Object value) {
-            return (AverageState) value;
-        }
-
-        @Override
-        public int compare(AverageState val1, AverageState val2) {
-            if (val1 == null) return -1;
-            return val1.compareTo(val2);
-        }
-
-        @Override
-        public AverageState readValueFrom(StreamInput in) throws IOException {
-            AverageState averageState = new AverageState();
-            averageState.sum = in.readDouble();
-            averageState.count = in.readVLong();
-            return averageState;
-        }
-
-        @Override
-        public void writeValueTo(StreamOutput out, AverageState v) throws IOException {
-            out.writeDouble(v.sum);
-            out.writeVLong(v.count);
-        }
-
-        @Override
-        public int fixedSize() {
-            return AVERAGE_STATE_SIZE;
         }
     }
 
     private final Signature signature;
     private final Signature boundSignature;
+    private final Supplier<AverageState> stateSupplier;
+    private final Supplier<FixedWidthType> stateDateTypeSupplier;
 
-    AverageAggregation(Signature signature, Signature boundSignature) {
+    AverageAggregation(Signature signature, Signature boundSignature, Supplier<AverageState> stateSupplier, Supplier<FixedWidthType> stateDateTypeSupplier) {
```

Comment:
Given that we have singleton instances for the type you could avoid wrapping it into a `Supplier`. Maybe also make it more concrete - given that there is a dependency that the type value must match the `AverageState`